src/Common/src/Microsoft/Win32/SafeHandles/NtlmSecuritySafeHandles.cs

```diff
@@ -0,0 +1,271 @@
+        {
+            //Message Confidentiality. Reference: https://msdn.microsoft.com/en-us/library/cc236707.aspx
+            Debug.Assert(_isSealingKey, "Cannot seal or unseal with signing key");
+            Debug.Assert(offset >= 0 && offset < buffer.Length, "Cannot sign with invalid offset " + offset);
+            Debug.Assert((count + offset) <= buffer.Length, "Cannot sign with invalid count ");
+
+            unsafe
+            {
+                fixed (byte* bytePtr = buffer)
+                {
+                    // Since RC4 is XOR-based, encrypt or decrypt is relative to input data
+                    // reference: https://msdn.microsoft.com/en-us/library/cc236707.aspx
+                    byte[] output = new byte[count];
+
+                    Interop.Crypto.EvpCipher(_cipherContext, output, (bytePtr + offset), count);
+                    return  output;
+                }
+            }
+        }
+
+        private static void MarshalUint(byte[] ptr, int offset, uint num)
+        {
+            for (int i = 0; i < 4; i++)
+            {
+                ptr[offset + i] = (byte) (num & 0xff);
+                num >>= 8;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a ntlm_type2*
+    /// </summary>
+    internal sealed class SafeNtlmType2Handle : SafeHandle
+    {
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurity.HeimNtlmFreeType2(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+
+        private SafeNtlmType2Handle() : base(IntPtr.Zero, true)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a ntlm_type3*
+    /// </summary>
+    internal sealed class SafeNtlmType3Handle : SafeHandle
+    {
+        private SafeNtlmType2Handle _type2Handle;
+        public SafeNtlmType3Handle(byte[] type2Data, int offset, int count) : base(IntPtr.Zero, true)
+        {
+            int status = Interop.NetSecurity.HeimNtlmDecodeType2(type2Data, offset, count, out _type2Handle);
+            Interop.NetSecurity.HeimdalNtlmException.AssertOrThrowIfError("HeimNtlmDecodeType2 failed", status);
+        }
+
+        public override bool IsInvalid
+        {
+            get { return (null != _type2Handle) && !_type2Handle.IsInvalid; }
+        }
+
+        public byte[] GetResponse(uint flags, string username, string password, string domain,
+                                  out byte[] sessionKey)
+        {
+            // reference for NTLM response: https://msdn.microsoft.com/en-us/library/cc236700.aspx
+            sessionKey = null;
+            SafeNtlmBufferHandle key;
+            int keyLen;
+            int status = Interop.NetSecurity.HeimNtlmNtKey(password, out key, out keyLen);
+            Interop.NetSecurity.HeimdalNtlmException.AssertOrThrowIfError("HeimNtlmKey failed", status);
```

Comment:
Same question as earlier about disposing of the key even in the case of failure (assuming it can fail but still produce a non-null key)